File: crates/opanel-agent/src/actions/systemd.rs
//! Driving the services the panel owns, and no others.

use crate::platform::svc::{self, Status};
use crate::registry::{decode, Failure, Outcome, Registry, Validate};
use serde::{Deserialize, Serialize};
use serde_json::json;

/// The units the panel may touch.
///
/// An allowlist rather than a shape check: the agent runs as root, so
/// "restart any unit you can name" is a different grant from "restart the
/// web server", and only the second one is wanted here.
const MANAGED_UNITS: &[&str] = &[
    "lshttpd",
    "lsws",
    "openlitespeed",
    "mariadb",
    "valkey",
    "nftables",
    "opanel-api",
    "opanel-agent",
    "sshd",
    "crond",
    "clamd@scan",
];

/// Spelled the way the Go agent formats the same list, so a caller that
/// matched on the message keeps working across the port.
const UNIT_VERBS: &str = "[start stop restart reload reload-or-restart]";

fn unit_allowed(name: &str) -> bool {
    MANAGED_UNITS.contains(&name)
}

#[derive(Debug, Default, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct UnitRequest {
    pub unit: String,
}

impl Validate for UnitRequest {
    fn validate(&self) -> Result<(), String> {
        if !svc::valid_unit(&self.unit) {
            return Err(format!("unit {:?} is not a valid unit name", self.unit));
        }
        if !unit_allowed(&self.unit) {
            return Err(format!("unit {:?} is not managed by opanel", self.unit));
        }
        Ok(())
    }
}

#[derive(Debug, Default, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct UnitActionRequest {
    pub unit: String,
    pub action: String,
}

impl Validate for UnitActionRequest {
    fn validate(&self) -> Result<(), String> {
        UnitRequest { unit: self.unit.clone() }.validate()?;
        match self.action.as_str() {
            "start" | "stop" | "restart" | "reload" | "reload-or-restart" => Ok(()),
            other => Err(format!("action {other:?} is not one of {UNIT_VERBS}")),
        }
    }
}

#[derive(Debug, Default, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct JournalRequest {
    pub unit: String,
    #[serde(default)]
    pub lines: i64,
}

impl Validate for JournalRequest {
    fn validate(&self) -> Result<(), String> {
        UnitRequest { unit: self.unit.clone() }.validate()
    }
}

#[derive(Serialize)]
struct StatusList {
    units: Vec<Status>,
}

pub fn register(r: &mut Registry) {
    r.register("systemd.status", 1, |payload| -> Outcome {
        let req: UnitRequest = decode(payload)?;
        let status = svc::get(&req.unit)?;
        serde_json::to_value(status).map_err(|e| Failure::internal(e.to_string()))
    });

    r.register("systemd.list", 1, |_| -> Outcome {
        let mut units = Vec::with_capacity(MANAGED_UNITS.len());
        for name in MANAGED_UNITS {
            let Ok(status) = svc::get(name) else { continue };
            // A unit that is neither installed nor running is not a service
            // in an unhappy state, it is a service this host does not have:
            // valkey on a box without it, clamd@scan before the scanner is
            // installed. Listing it would be a permanent red row nobody can
            // fix.
            if status.enabled == "not-found" && !status.running() {
                continue;
            }
            units.push(status);
        }
        serde_json::to_value(StatusList { units }).map_err(|e| Failure::internal(e.to_string()))
    });

    r.register("systemd.action", 1, |payload| -> Outcome {
        let req: UnitActionRequest = decode(payload)?;
        match req.action.as_str() {
            "start" => svc::start(&req.unit),
            "stop" => svc::stop(&req.unit),
            "restart" => svc::restart(&req.unit),
            "reload" => svc::reload(&req.unit),
            _ => svc::reload_or_restart(&req.unit),
        }?;
        let status = svc::get(&req.unit)?;
        serde_json::to_value(status).map_err(|e| Failure::internal(e.to_string()))
    });

    r.register("systemd.journal", 1, |payload| -> Outcome {
        let req: JournalRequest = decode(payload)?;
        let text = svc::journal(&req.unit, req.lines)?;
        Ok(json!({ "text": text }))
    });
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_managed_units_pass_validation() {
        assert!(UnitRequest { unit: "mariadb".into() }.validate().is_ok());
        // Well-formed, real, and refused: the allowlist is what stops a flaw
        // anywhere in the panel from reaching systemd-journald or auditd.
        let err = UnitRequest { unit: "systemd-journald".into() }.validate().unwrap_err();
        assert!(err.contains("not managed by opanel"), "{err}");
        let err = UnitRequest { unit: "--now".into() }.validate().unwrap_err();
        assert!(err.contains("not a valid unit name"), "{err}");
    }

    #[test]
    fn verbs_are_the_five_the_panel_needs() {
        for verb in ["start", "stop", "restart", "reload", "reload-or-restart"] {
            let req = UnitActionRequest { unit: "crond".into(), action: verb.into() };
            assert!(req.validate().is_ok(), "{verb} should be accepted");
        }
        let req = UnitActionRequest { unit: "crond".into(), action: "mask".into() };
        assert!(req.validate().is_err(), "mask is not one of the five");
    }
}

File: internal/agent/actions/systemd.go
package actions

import (
	"context"
	"fmt"
	"slices"
	"strings"

	"github.com/bnixvn/opanel-ent/internal/agent"
	"github.com/bnixvn/opanel-ent/internal/phpmgr"
	"github.com/bnixvn/opanel-ent/internal/platform/svc"
)

// managedUnits is the allowlist of units the panel may control.
//
// The unix socket already restricts *who* can call; this restricts *what*
// they can reach. Without it, a flaw anywhere in the API would let a caller
// stop sshd or systemd-journald through a perfectly well-formed request.
var managedUnits = []string{
	"httpd",   // Apache
	"lshttpd", // LiteSpeed Enterprise, when a licence is installed
	"mariadb",
	"valkey",
	"nftables",
	"opanel-api",
	"opanel-agent",
	"sshd",
	"crond",
	"clamd@scan",
	// One pool manager per PHP version. Listed rather than discovered
	// because the allowlist is the point: it is what stops a flaw anywhere
	// in the panel from reaching a unit nobody meant it to touch. A version
	// that is not installed is skipped by systemd.list, so this costs
	// nothing on a host that runs one of them.
	"php74-php-fpm",
	"php80-php-fpm",
	"php81-php-fpm",
	"php82-php-fpm",
	"php83-php-fpm",
	"php84-php-fpm",
	"php85-php-fpm",
}

func unitAllowed(name string) bool { return slices.Contains(managedUnits, name) }

// UnitRequest names a unit to inspect or act on.
type UnitRequest struct {
	Unit string `json:"unit"`
}

// Validate rejects malformed and unmanaged unit names.
func (r *UnitRequest) Validate() error {
	if !svc.ValidUnit(r.Unit) {
		return fmt.Errorf("unit %q is not a valid unit name", r.Unit)
	}
	if !unitAllowed(r.Unit) {
		return fmt.Errorf("unit %q is not managed by opanel", r.Unit)
	}
	return nil
}

// UnitActionRequest asks for a state change.
type UnitActionRequest struct {
	Unit   string `json:"unit"`
	Action string `json:"action"` // start|stop|restart|reload|reload-or-restart
}

var unitVerbs = []string{"start", "stop", "restart", "reload", "reload-or-restart"}

// Validate checks both the unit and the verb.
func (r *UnitActionRequest) Validate() error {
	if err := (&UnitRequest{Unit: r.Unit}).Validate(); err != nil {
		return err
	}
	if !slices.Contains(unitVerbs, r.Action) {
		return fmt.Errorf("action %q is not one of %v", r.Action, unitVerbs)
	}
	return nil
}

// JournalRequest asks for recent log lines.
type JournalRequest struct {
	Unit  string `json:"unit"`
	Lines int    `json:"lines"`
}

// Validate checks the unit and clamps the line count.
func (r *JournalRequest) Validate() error {
	if err := (&UnitRequest{Unit: r.Unit}).Validate(); err != nil {
		return err
	}
	if r.Lines <= 0 || r.Lines > 5000 {
		r.Lines = 200
	}
	return nil
}

// JournalResult carries log text.
type JournalResult struct {
	Text string `json:"text"`
}

// UnitListResult reports the status of every managed unit.
type UnitListResult struct {
	Units []svc.Status `json:"units"`
}

// idleFPM reports a pool manager that is stopped because no site asks for
// that PHP version.
func idleFPM(unit string, st svc.Status) bool {
	version, ok := fpmUnitVersion(unit)
	if !ok || st.Running() {
		return false
	}
	return phpmgr.NewRemi().PoolCount(version) == 0
}

// fpmUnitVersion turns "php84-php-fpm" back into "8.4".
func fpmUnitVersion(unit string) (string, bool) {
	rest, ok := strings.CutSuffix(unit, "-php-fpm")
	if !ok {
		return "", false
	}
	digits, ok := strings.CutPrefix(rest, "php")
	if !ok || len(digits) != 2 {
		return "", false
	}
	return digits[:1] + "." + digits[1:], true
}

func registerSystemd(r *agent.Registry) {
	agent.Register(r, "systemd.status", 1, func(ctx context.Context, in UnitRequest) (svc.Status, error) {
		return svc.Get(ctx, in.Unit)
	})

	agent.Register(r, "systemd.list", 1, func(ctx context.Context, _ struct{}) (UnitListResult, error) {
		out := UnitListResult{Units: make([]svc.Status, 0, len(managedUnits))}
		for _, u := range managedUnits {
			st, err := svc.Get(ctx, u)
			if err != nil {
				continue
			}
			// Units that are not installed are noise on a fresh host.
			if st.Enabled == "not-found" && st.Active != "active" {
				continue
			}
			// A PHP version nobody's site uses is deliberately stopped, and
			// showing it here as a stopped service reads as something to go
			// and fix. The PHP page is where a version's state belongs,
			// alongside the number of sites that explains it.
			if idleFPM(u, st) {
				continue
			}
			out.Units = append(out.Units, st)
		}
		return out, nil
	})

	agent.Register(r, "systemd.action", 1, func(ctx context.Context, in UnitActionRequest) (svc.Status, error) {
		var err error
		switch in.Action {
		case "start":
			err = svc.Start(ctx, in.Unit)
		case "stop":
			err = svc.Stop(ctx, in.Unit)
		case "restart":
			err = svc.Restart(ctx, in.Unit)
		case "reload":
			err = svc.Reload(ctx, in.Unit)
		case "reload-or-restart":
			err = svc.ReloadOrRestart(ctx, in.Unit)
		}
		if err != nil {
			return svc.Status{}, err
		}
		return svc.Get(ctx, in.Unit)
	})

	agent.Register(r, "systemd.journal", 1, func(ctx context.Context, in JournalRequest) (JournalResult, error) {
		text, err := svc.Journal(ctx, in.Unit, in.Lines)
		return JournalResult{Text: text}, err
	})
}
